### PR TITLE
Generalize Build Scan CSV parsing and Java invocation logic

### PR DESCRIPTION
**Pull request by erichaagdev**
# Testing

### Experiment 2

```
./02-validate-local-build-caching-same-location.sh \
  --git-repo 'https://github.com/etiennestuder/java-ordered-properties' \
  --git-branch ge \
  --tasks build
```

image

### Experiment 3

```
./03-validate-local-build-caching-different-locations.sh \
  --git-repo 'https://github.com/etiennestuder/java-ordered-properties' \
  --git-branch ge \
  --tasks build 
```

image

### Experiment 4

```
./04-validate-remote-build-caching-ci-ci.sh \
  -1 'https://ge.solutions-team.gradle.com/s/6x7tulbxvdt66' \
  -2 'https://ge.solutions-team.gradle.com/s/xm7udfl7jp4hs'
```

image

### Experiment 5

```
./05-validate-remote-build-caching-ci-local.sh \
  -1 'https://ge.solutions-team.gradle.com/s/6x7tulbxvdt66'
```

image

### Experiment 1

```
./01-validate-local-build-caching-same-location.sh \
  --git-repo 'git@github.com:erichaagdev/openapi-generator.git' \
  --git-branch gradle-enterprise \
  --goals package \
  --args '-DskipTests'
```

image

### Experiment 2

```
./02-validate-local-build-caching-different-locations.sh \
  --git-repo 'git@github.com:erichaagdev/openapi-generator.git' \
  --git-branch gradle-enterprise \
  --goals package \
  --args '-DskipTests'
```

image

### Experiment 3

```
./03-validate-remote-build-caching-ci-ci.sh \
  -1 'https://ge.solutions-team.gradle.com/s/ed6a3h7m3swpa' \
  -2 'https://ge.solutions-team.gradle.com/s/eq5u5tbq6xdyo'
```

image

### Experiment 4

```
./04-validate-remote-build-caching-ci-local.sh\
  -1 'https://ge.solutions-team.gradle.com/s/ed6a3h7m3swpa' \
  --args '-DskipTests'
```

image

### CODE REVIEW

File: components/scripts/lib/java.sh [hunk 1]
```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env bash
+
+invoke_java() {
```
Comment (jthurne) [inline components/scripts/lib/java.sh:3]: 0: This isn't something to do right now, but I had originally borrowed this code from the launch scripts that the Gradle Application Plugin generates.  But it has a number of problems.  The BT Team recently updated the application plugin to generate a much better launch script. At some point, we should update this code to be based on that.

Again, this is good for now though.

File: components/scripts/lib/libs.sh [hunk 2]
```diff
@@ -12,41 +12,47 @@ failed_to_load_lib() {
...
 }
 
+# shellcheck source=lib/build_scan.sh
+source "${LIB_DIR}/build_scan.sh" || failed_to_load_lib build_scan.sh
+
 # shellcheck source=lib/color.sh
 source "${LIB_DIR}/color.sh" || failed_to_load_lib color.sh
 
-# shellcheck source=lib/exit-code.sh
-source "${LIB_DIR}/exit-code.sh" || failed_to_load_lib exit-code.sh
```
Comment (jthurne) [inline components/scripts/lib/libs.sh:19]: 0: The order here may be important. I had this loaded early on because the other library scripts depend on the functions in this one.  But since everything is loaded before the real script logic is executed, maybe it's okay.  Mentioning it in case we run into problems (this would be something to look at as a possible cause).

File: components/scripts/lib/parse.sh [hunk 3]
```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env bash
```
Comment (jthurne) [inline components/scripts/lib/parse.sh:1]: -1: I recommend naming this `build-scan-parse.sh` so that it's clear this is still focused on build scans.

File: components/scripts/lib/parse.sh [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+     avoided_from_cache_num_tasks[idx]="${field_14}"
+     avoided_from_cache_avoidance_savings[idx]="${field_15}"
+     executed_cacheable_num_tasks[idx]="${field_16}"
+     executed_cacheable_duration[idx]="${field_17}"
+     executed_not_cacheable_num_tasks[idx]="${field_18}"
+     executed_not_cacheable_duration[idx]="${field_19}"
+
+     ((idx++))
+  done <<< "${build_scan_csv}"
+}
```
Comment (jthurne) [inline components/scripts/lib/parse.sh:77]: 0: Missing newline at the end of the file

**Review by jthurne — APPROVED** [review]
Just a minor naming suggestion, otherwise looks good to me.

File: components/scripts/lib/libs.sh (hunk 2 above)
Comment (erichaagdev) [reply components/scripts/lib/libs.sh:19]: I'm running through some tests now to verify functionality. Do you want me to leave this ordering if everything looks normal?

Comment (jthurne) [reply components/scripts/lib/libs.sh:19]: Yes. If we can get away with the new ordering, that would be great. It's definitely an improvement.

Comment (erichaagdev) [reply components/scripts/lib/libs.sh:19]: Reordering the files did not seem to have any negative impact based on the tests I ran. I'm not sure if there are any edge cases to consider.

File: components/scripts/lib/parse.sh (hunk 3 above)
Comment (erichaagdev) [reply components/scripts/lib/parse.sh]: Addressed by fa178a1c3b65193c584d280eea252be392953871

File: components/scripts/lib/parse.sh (hunk 4 above)
Comment (erichaagdev) [reply components/scripts/lib/parse.sh]: Addressed by fa178a1c3b65193c584d280eea252be392953871

File: components/scripts/lib/libs.sh (hunk 2 above)
Comment (jthurne) [reply components/scripts/lib/libs.sh:19]: Let's go with it for now.

File: components/scripts/lib/java.sh (hunk 1 above)
Comment (etiennestuder) [reply components/scripts/lib/java.sh:3]: I suggest to create a GH issue for this update.

Comment (jthurne) [reply components/scripts/lib/java.sh:3]: Issue created: https://github.com/gradle/gradle-enterprise-build-validation-scripts/issues/176
